controllers/tekton/pipeline/pipeline_controller.go

```diff
@@ -0,0 +1,195 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pipeline
+
+import (
+	"context"
+	"fmt"
+
+	tektonv1 "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	devopsv2alpha1 "kubesphere.io/devops/pkg/api/devops/v2alpha1"
+)
+
+// PipelineReconciler reconciles a Pipeline object
+type PipelineReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=devops.kubesphere.io,resources=pipelines,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=devops.kubesphere.io,resources=pipelines/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=devops.kubesphere.io,resources=pipelines/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the Pipeline object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+func (r *PipelineReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+
+	klog.Info(fmt.Sprintf("req: %v", req))
+
+	// get cr by its name
+	pipeline := &devopsv2alpha1.Pipeline{}
+	if err := r.Get(ctx, req.NamespacedName, pipeline); err != nil {
+		klog.Error(err, "unable to fetch pipeline crd resources")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// translate our pipeline CRD to tekton CRDs, e.g. Task, Pipeline, PipelineResource
+	if err := r.reconcileTektonCrd(ctx, req.Namespace, pipeline); err != nil {
+		klog.Error(err, "unable to reconcile tekton crd resources")
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *PipelineReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&devopsv2alpha1.Pipeline{}).
+		Complete(r)
+}
+
+// reconcileTektonCrd translates our Pipeline CRD to Tekton CRDs
+func (r *PipelineReconciler) reconcileTektonCrd(ctx context.Context, namespace string, pipeline *devopsv2alpha1.Pipeline) error {
+	// print the pipeline name and the number of its tasks.
+	klog.Info(fmt.Sprintf("Pipeline name: %s\tTask num: %d", pipeline.Name, len(pipeline.Spec.Tasks)))
```

Comment:
same suggestion

Reply:
> same suggestion

I would like to do the same revision in this [comment](https://github.com/kubesphere/ks-devops/pull/179#discussion_r689218276).